### PR TITLE
[WIP] Linked datasets

### LINKED ISSUE #7: Linked datasets field
Original spec
> Links to other RIDL datasets; user can pick one or many datasets that are related to this one (upd)

Implementation:

This will need a custom preset on the unhcr extension:
* Create `presets.json` in the `schemas` folder, add it to the relevant configuration option
* Create a `linked_datasets` preset with the following items:
  * Form snippet. There are two options:
     1. Prefilled list that allows multiple selection with the dataset titles (from the organizations the user belongs to). This has the problem of a very long list if the user belongs to many organizations or is a sysadmin (which can see all orgs)
    2. Search widget that queries the search API, filtering by org and renders the results, which can be then selected. We will need a custom js module for this [1].
  * Validator [2] : 
    1. Get all organizations (data containers) the user belongs to
    2. For each dataset id provided, check if the dataset belongs to one of the above
    3. If not raise `Invalid`
  * Display snippet: It will include a helper function [3] that translates the stored dataset ids into dataset titles (querying `package_search` or `package_show`, and creates a list of links.

[1] Modules are jQuery classes like [this one](https://github.com/frictionlessdata/ckanext-validation/blob/master/ckanext/validation/fanstatic/js/module-modal-dialog.js) and they are called from [templates](https://github.com/frictionlessdata/ckanext-validation/blob/master/ckanext/validation/templates/validation/snippets/validation_report_dialog.html#L11)
[2] Validators are created in a `validators.py` and registered with `IValidators`, check ckanext-validation [here](https://github.com/frictionlessdata/ckanext-validation/blob/master/ckanext/validation/plugin.py#L231) and [here](https://github.com/frictionlessdata/ckanext-validation/blob/master/ckanext/validation/validators.py#L11))
[3] Helper functions are functions that can be called from templates. They are registered with `ITemplateHelpers` like [here](https://github.com/frictionlessdata/ckanext-validation/blob/master/ckanext/validation/plugin.py#L105)

Estimate: 3-4 days

### PR DESCRIPTION
**Pull request by roll**
- fixes #7

---

# Tasks

- [x] optimize linked datasets preparation speed
- [x] swith to UI component for seachable multiple select (don't need to resolve macro form.select issue)
- [x] display_snippet's links are ugly (see `url_for` uses `id` instead of `name`)
- [x] create an issue for backward links for linked datasets 
- [x] how to sort datasets for form/display (`select2` groupping?)
- [x] tests, fix travis build

### CODE REVIEW

File: ckanext/unhcr/templates/scheming/form_snippets/linked_datasets.html [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{% import 'macros/form.html' as form %}
+
+<div>
+  {{ form.select(
+      'linked_datasets',
+      label='Linked Datasets',
+      id='field-linked-datasets',
+      options=h.get_linked_datasets_options(exclude_id=data.get('id')),
+      selected=data[field.field_name],
```
Comment (roll) [inline ckanext/unhcr/templates/scheming/form_snippets/linked_datasets.html]: @amercader 
Do we have a macros with the multiple select support?

File: ckanext/unhcr/templates/scheming/form_snippets/linked_datasets.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+      attrs={'multiple': 'multiple'},
+      classes=['control-full']
+      )
+   }}
+</div>
+
+<div id="test" data-module="linked-datasets"></div>
```
Comment (roll) [inline ckanext/unhcr/templates/scheming/form_snippets/linked_datasets.html]: @amercader 
Can't we re-use a form input component e.g. used for tags? If we need to solve this `This has the problem of a very long list if the user belongs to many organizations or is a sysadmin (which can see all orgs)`

**Comment by amercader** [comment]
@roll I've been playing with this for a while and the current helper is just too slow when you are an admin and can access all data-containers. This change uses the search to get all datasets and it's much faster:

```diff
diff --git a/ckanext/unhcr/helpers.py b/ckanext/unhcr/helpers.py
index 2c225f0..b0e0377 100644
--- a/ckanext/unhcr/helpers.py
+++ b/ckanext/unhcr/helpers.py
@@ -61,14 +61,21 @@ def get_linked_datasets_for_form(exclude_id=None):
     # Get datasets
     datasets = []
     get_containers = toolkit.get_action('organization_list_for_user')
-    get_container = toolkit.get_action('organization_show')
+    dataset_search = toolkit.get_action('package_search')
     containers = get_containers(context, {'id': toolkit.c.userobj.id})
+    fq_list = []
     for container in containers:
-        container = get_container(context, {'id': container['id'], 'include_datasets': True})
-        for package in container['packages']:
-            if package['id'] == exclude_id:
-                continue
-            datasets.append({'text': package['title'], 'value': package['id']})
+        fq_list.append('owner_org:{}'.format(container['id']))
+    search = dataset_search(context, {
+        'fq': ' OR '.join(fq_list),
+        'include_private': True,
+        }
+    )
+
+    for package in search['results']:
+        if package['id'] == exclude_id:
+            continue
+        datasets.append({'text': package['title'], 'value': package['id']})


```

**Comment by roll** [comment]
@amercader 
Thanks awesome.

**Comment by roll** [comment]
@amercader 
Would you like to play with the updated form component? 

Also, I track not resolved things in the first message.

**Comment by amercader** [comment]
Great work @roll. I improved the ordering of the options by grouping by Data containers:

screenshot from 2018-04-24 13-55-31

**Pull request merged**